### DISCUSSION TITLE
Things you can do with Tabby

### DISCUSSION THREAD

**Discussion by Eugeny** [Eugeny/tabby#5259] [FAQ & Tips]
_This is a small collection of random things that you might have missed since they are not as easily discoverable_

## General

* *Text editing*: In most shells and editors, you can use `Ctrl-Arrows` (`⌥-Arrows`) and `Home`/`End` (`⌘-Left`/`⌘-Right`) to move the cursor around in the prompt.
missing in the main font only.
* *Intelligent Ctrl-C*: When text is selected, Ctrl-C will copy it on the first press, and issue an actual abort on the second press. You can disable this by unassigning the _Intelligent Ctrl-C_ hotkey.
* *Quake mode / docking*: you can make the main terminal window show up / hide with a hotkey and have it stick to the top part of screen by enabling _docking_ in Settings -> Window.
* *Fallback fonts*: Your favourite font has no icons or [Powerline](https://github.com/powerline/powerline) symbols? You can set a Powerline-enabled font as the _fallback font_ in `Settings -> Appearance` and Tabby will use it for the symbols 

## Split panes

* *Type into all panes at once* (also known as "broadcast"): You can type into all panes by pressing `Ctrl-Shift-I` (`⌘-Shift-i`). Click any pane to cancel focus.
* *Pane rearrange hotkey*: hold `Ctrl-Shift` (`⌘-Shift`) and pane drag handles will appear. You can rearrange panes or drag them out as new tabs: ![](https://user-images.githubusercontent.com/161476/147154268-8d98b429-d6bb-4594-a280-4c24818b491f.png)
* *Maximizing panes*: `Ctrl-Alt-Enter` (`⌘-⌥-Enter`) to make the current pane temporarily full-size, press again to cancel.
* *Keyboard navigation around panes*:  you can navigate around the splits with `Ctrl-Alt-Arrows` (`⌘-⌥-Arrows`)
* *Saving pane layouts*: with multiple panes present, right-click a tab and select `Save layout as profile`. You can even combine different pane types (local/SSH/serial) by opening multiple tabs first and then dragging them one into another to combine, or by using the `Switch pane profile` context menu item or shortcut.
* *Exploding and combining tabs*:  `Ctrl-Shift-.` (`⌘-Shift-.`) to turn all split panes in the current tab into their own tabs / `Ctrl-Shift-,` (`⌘-Shift-,`) to combine all tabs back into the current one as panes. Combine this with _broadcast_ (`Ctrl-Shift-i` / `⌘-Shift-i`) to quickly send commands to all tabs! `1.0.173+`

## Networking

* *Debugging socket servers*: you can use the `Raw socket connection` profile template to debug various servers, `nc` style. Combine with `Hexadecimal` input and output modes in the profile settings to work with binary protocols: 

<img width="300" src="https://user-images.githubusercontent.com/161476/147154145-50746f87-7a01-483f-9fd7-624a0806a31c.png"/>

* *SOCKS proxy support for SSH*: You can connect to an SSH host via a proxy without having to set up a ProxyCommand: 

<img width="300" src="https://user-images.githubusercontent.com/161476/147152121-10991885-595d-4697-a081-55b152933967.png"/>

## Profiles

* *Changing defaults for all profiles*: You can edit the profile defaults for each profile type (local/SSH/serial) in Settings -> Profiles -> Advanced.
* *Separate profile selectors*: you can assign a hotkey to show a profile selector just for SSH profiles / serial profiles etc - see the `profile-selectors.*` hotkeys (not assigned by default).

## Other
* *Docker*: there is an official Docker plugin that lets you spawn shells inside new or existing containers.
* *Working with buggy serial adapters*: the `Slow feed` option in the profile's advanced settings will slow down output to the serial driver by sending content byte by byte.
* *Activity notification*: select `Notify on activity` in a tab's context menu to get a notification when there's new output.
* *Process completion notification*: select `Notify when done` in a tab's context menu to get a notification when a child process completes in the tab (only works for local shells): 
 
<img width="300" src="https://user-images.githubusercontent.com/161476/147154434-0262c60e-d0f5-4e7d-818d-a4b66cd73db4.png"/>

**Comment by cburkins** [comment]
Silly question, the "panes" feature mentioned above is intriguing to me, but I can't figure out how to trigger/configure it.   The directions above for panes seems to be relevant when you already have multiple panes open.   How do I go about getting two tabs into a split pane configuration

**Reply by cburkins** [reply]
I'm on a Mac, not sure if that's relevant.

**Reply by Eugeny** [reply]
You can either split an existing tab into two (context menu -> split or the split hotkeys), or drag another tab onto it.

**Reply by lino-man** [reply]
> Silly question, the "panes" feature mentioned above is intriguing to me, but I can't figure out how to trigger/configure it. The directions above for panes seems to be relevant when you already have multiple panes open. How do I go about getting two tabs into a split pane configuration

Just the same here. Just cant.. And on context menu does not have the  split option. Im on windows 10

**Reply by Eugeny** [reply]
Just to clarify, I mean this: 
![image](https://user-images.githubusercontent.com/161476/152762078-fd52d3ae-5b3c-4f8f-8931-173b4a8dff19.png)

**Reply by cburkins** [reply]
Ah, helpful screenshot !!  The "split" command is on the context menu on the "body" of the tab, not on the tab itself, which apparently is where I was expecting to find it.   So... my use case is that I want to connect to two different remote SSH hosts via two panes in the same tab.   I think I would bring up a new local shell using the "bash" profile, like this:

Then I would context-click (i.e. right-click) in the body of each side of the pane, and select "switch profile" to connect to the desired remote machine for that pane.   

Am I thinking about this correctly ?

![image](https://user-images.githubusercontent.com/9342308/152795920-c40432bb-459f-4347-a570-d6fe1cbca965.png)

**Reply by Eugeny** [reply]
Yes, that's one way to go about it. Alternatively, you can connect to the first host in the first tab, connect to the 2nd host in the 2nd tab, and then drag the 1st tab onto the side of the 2nd tab.

**Reply by cburkins** [reply]
Ah.  That would be my preferred workflow.    I didn't think to drag onto the top of another tab, but that makes sense now.   It worked two or three times for me, but now I can't get it to work again.    There are also "drop zones" at the top, bottom, right, left, but when I drop onto one of those, it seems to kill the tab altogether.   Trying to figure out how I can show you.....    Maybe I can make an animated gif that will upload ?

**Reply by Eugeny** [reply]
Make sure you're not actually dragging the tab onto itself - it's a known bug.

**Reply by cburkins** [reply]
https://user-images.githubusercontent.com/9342308/152894290-137cc2a9-65ec-486c-94f5-4d173beeb952.mp4

Quick demo of my.... confusion, I guess.
FYI, video has sound (me narrating), toggle it on when you play it

**Reply by cburkins** [reply]
OS is MacOS 11.6.1
Tabby version is 1.0.170 with experimental features enabled.

**Reply by rushowr** [reply]
1. Tabs will indeed be moved around the order if you drag them along the
tab-bar.

2. Dropping a tab onto a drop zone within itself won't work..

3. Dropping a DIFFERENT tab onto a drop zone in your current tab, should
split the tab into two sub tabs.

4. The drop zones (this is in response to your inner
dialogue/commentary)  represent the placement of the dropped tab and
where the current tab will go in relation to that drop. I.e. the top
zone will create a horizontally split pane set, with the dropped tab in
the upper position. Make sense?

Any mistakes in my answer can be blamed on the fact that I'm a bit blunt
and even though I've not used the feature you're talking about, I have
worked with tabs and splitting windows/tabs in other applications, and
they all followed the same principles as the ones I've used in my
answer. Also, the dropping the tab onto itself wouldn't make logical
sense (to me) to try unless completely out of logical options to try,
AND it was stated earlier in the thread, so I knew I had the right
answer to that one ;)

Anyone out there see an error in my answer? I absolutely welcome corrections of factual information, and will gladly correct and give full acknowledgement of the correction author.


Note:
(opinions on my style of answering are welcome but likely will not result in changing my ways ;-)
I'm a 44 year-old schizophrenic high school dropout who has never been to college (obviously, *haha*) but rather
figured out about 95% of things on my own (in multiple IT related fields) and ended up a well paid and decently in demand consultant, not to mention a raving jerk who can usually at least back up his point :). I enjoy a little friendly sh**-giving here and there.


Cheers

**Reply by Eugeny** [reply]
@cburkins I see why it's so confusing now! You don't need `Cmd-Shift` to just merge two tabs together.
* With tab03 active, start dragging tab02 by its "index" 
* Drop it onto one of the drop zones

**Reply by cburkins** [reply]
Ah, never thought of that !   And took me a second to understand "by index".   Dragging via the tab name vs the tab number, right ?  Works great !      (though I think removing the index number might be a configurable option, wonder how that might affect usability here).  Either way, super big thanks for the new feature !!!!

**Reply by cburkins** [reply]
> 1. Tabs will indeed be moved around the order if you drag them along the
>    tab-bar.
> 2. Dropping a tab onto a drop zone within itself won't work..
> 3. Dropping a DIFFERENT tab onto a drop zone in your current tab, should
>    split the tab into two sub tabs.
> 4. The drop zones (this is in response to your inner
>    dialogue/commentary)  represent the placement of the dropped tab and
>    where the current tab will go in relation to that drop. I.e. the top
>    zone will create a horizontally split pane set, with the dropped tab in
>    the upper position. Make sense?
> 
> Any mistakes in my answer can be blamed on the fact that I'm a bit blunt and even though I've not used the feature you're talking about, I have worked with tabs and splitting windows/tabs in other applications, and they all followed the same principles as the ones I've used in my answer. Also, the dropping the tab onto itself wouldn't make logical sense (to me) to try unless completely out of logical options to try, AND it was stated earlier in the thread, so I knew I had the right answer to that one ;)
> 
> Anyone out there see an error in my answer? I absolutely welcome corrections of factual information, and will gladly correct and give full acknowledgement of the correction author.
> 
> Note: (opinions on my style of answering are welcome but likely will not result in changing my ways ;-) I'm a 44 year-old schizophrenic high school dropout who has never been to college (obviously, _haha_) but rather figured out about 95% of things on my own (in multiple IT related fields) and ended up a well paid and decently in demand consultant, not to mention a raving jerk who can usually at least back up his point :). I enjoy a little friendly sh**-giving here and there.
> 
> Cheers

Thanks for the thoughts, very helpful.

**Comment by aesrys** [comment]
Is there some kind of timeout on the pane rearrange hotkey? I can see the pane labels if I hold down the keys but they only stay visible for a couple of seconds.

**Reply by Eugeny** [reply]
This should be fixed in the nightly build now.

**Reply by aesrys** [reply]
Thanks, appreciate the quick response.

**Comment by siu91** [comment]
LGTM

**Comment by gkso** [comment]
Can we have "notify when done" for remote tabs? pls~

**Reply by Eugeny** [reply]
Unfortunately not since SSH provides no information on the processes running in a session. This could be a plugin idea with corresponding shell plugin on the other side

**Reply by gkso** [reply]
OK. My idea is that after the remote task is done, it writes some output to SSH. And Tabby receives the output and send a notification to user. Could you let me know how to write a plugin like this? Thanks.

**Reply by Eugeny** [reply]
Start with the demo plugin (tabby-clippy) - you'll need a [TabContextMenuItemProvider](https://docs.tabby.sh/classes/TabContextMenuItemProvider.html) to add the menu item and a [SessionMiddleware](https://docs.tabby.sh/terminal/classes/SessionMiddleware.html) to listen for session output and react to some custom escape sequence to signal that the command is done (you could add this sequence to $PS1)

**Comment by kktt007** [comment]
> <img alt="" width="300" src="https://user-images.githubusercontent.com/161476/147154268-8d98b429-d6bb-4594-a280-4c24818b491f.png">

May i know your theme or dot file, looks nice. Thanks.

**Reply by Eugeny** [reply]
I don't remember where I got it from, but here are the colours:

```yaml
    - name: Custom theme
      foreground: '#9cc2c3'
      background: '#001E27'
      cursor: '#f34b00'
      colors:
        - '#002831'
        - '#d11c24'
        - '#6cbe6c'
        - '#a57706'
        - '#2176c7'
        - '#c61c6f'
        - '#259286'
        - '#eae3cb'
        - '#006488'
        - '#f5163b'
        - '#51ef84'
        - '#b27e28'
        - '#178ec8'
        - '#e24d8e'
        - '#00b39e'
        - '#fcf4dc'
      cursorAccent: '#ded8df'
      selection: '#f4511e'
      selectionForeground: '#ffffff'
```

**Comment by L1uTongwei** [comment]
Well, how to hide my tabs to make it look like a real fullscreen?

I prefer a real tty just like you press Ctrl + Fx in Linux

**Reply by L1uTongwei** [reply]
Or just like you press F11 in Windows Terminal

**Comment by AndreiaMachado** [comment]
Its possible to save in a profile two or more split panes? Right now I can save the layout but I can't specified the folders that profile can open.

**Reply by Eugeny** [reply]
Not directly, but you can set up a split tab exactly the way you want it to be (including panes with different profiles), and then save - it will remember the pane-specific settings as well.

**Reply by AndreiaMachado** [reply]
ok. thank you for the tip.

**Comment by DarkestRainn** [comment]
Hi. Is it possible to make it so that when you right-click on a program icon in the taskbar, a menu appears with quick launch of different profiles, like for example the new window function (or new incognito window) in browsers?
